refactor: cleaning up macros::functions::utils

benesjan opened the pull request:
This PR was quite a journey:

## Step 1

When trying to figure out how to rename the incorrectly named `get_fn_visibility` in a PR down the stack I realized that it's better to just drop it because:
- it was only used by `create_assert_correct_initializer_args`, `create_mark_as_initialized` and `create_init_check` functions which in turn where only called in `transfrom_private` and `transform_public`,
- these functions contained only 2 lines of code.

Because of this ^ it's better to just directly inline contents of these funcs into the `transform_*` functions.

When I finished this I checked where are the is `is_fn_private` and `is_fn_public` utils used which led me to step 2:

## Step 2
When checking when are the above mentioned functions used I realized that we use them in `create_authorize_once_check` and in the `stub_fn`. In both of these cases the code was just terrible because these functions are called from `transform_private` and `transform_public` functions hence the flow was:

`transform_private` --> `stub_fn` --> are we in private? --> if yes call `create_private_stub`

which I could replace with:

`transform_private` --> `create_private_stub`

So I dropped the `stub_fn` and just called the `create_private_stub` directly.

The similar was the case for `create_authorize_once_check` where I could simply pass in an `is_private` arg to the function and get rid of the `is_fn_private` dependency.

## Step 3
Once I did this I realized that I don't need the `EXTERNAL_REGISTRY` I introduced in the PR down the stack at all, removing a lot of the sad ugly complexity I needed to introduce there 🎉


## Future work
- The whole `noir-projects/aztec-nr/aztec/src/macros/functions` directory is very messy and I plan on move the code around into more logical places in the dir once this stack of PRs is merged.
- I think the `stub` name is non-descriptive enough so will try to come up with something better.

benesjan commented:
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17997** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17997?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17958** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17958?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17884** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17840** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17840?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17840?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17827** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17827?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -1,9 +1,13 @@
+// Instead of manually serializing arguments and creating call interfaces, stubs allow natural syntax, e.g. for
+// enqueuing calls to public functions:
+//
+//   ExternalContract.at(address).some_method(arg1, arg2).enqueue()
```
Updated docs here as it never described what the stubs actually are.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -320,42 +335,6 @@ pub(crate) comptime fn transform_utility(f: FunctionDefinition) {
     f.set_return_public(true);
 }
 
-comptime fn create_internal_check(f: FunctionDefinition) -> Quoted {
-    let name = f.name();
-    let assertion_message = f"Function {name} can only be called internally";
-    quote { assert(context.msg_sender().unwrap() == context.this_address(), $assertion_message); }
-}
-
-comptime fn create_view_check(f: FunctionDefinition) -> Quoted {
-    let name = f.name();
-    let assertion_message = f"Function {name} can only be called statically";
-    if is_fn_private(f) {
-        // Here `context` is of type context::PrivateContext
-        quote { assert(context.inputs.call_context.is_static_call, $assertion_message); }
-    } else {
-        // Here `context` is of type context::PublicContext
-        quote { assert(context.is_static_call(), $assertion_message); }
-    }
-}
-
-comptime fn create_assert_correct_initializer_args(f: FunctionDefinition) -> Quoted {
-    let fn_visibility = get_fn_visibility(f);
-    f"dep::aztec::macros::functions::initialization_utils::assert_initialization_matches_address_preimage_{fn_visibility}(context);"
-        .quoted_contents()
-}
-
-comptime fn create_mark_as_initialized(f: FunctionDefinition) -> Quoted {
-    let fn_visibility = get_fn_visibility(f);
-    f"dep::aztec::macros::functions::initialization_utils::mark_as_initialized_{fn_visibility}(&mut context);"
-        .quoted_contents()
-}
-
-comptime fn create_init_check(f: FunctionDefinition) -> Quoted {
-    let fn_visibility = get_fn_visibility(f);
-    f"dep::aztec::macros::functions::initialization_utils::assert_is_initialized_{fn_visibility}(&mut context);"
-        .quoted_contents()
-}
```
Now I just inline all these directly which significantly decreased code complexity.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -19,24 +23,27 @@ comptime global FROM_FIELD: TypedExpr = {
 
-pub comptime fn stub_fn(f: FunctionDefinition) -> Quoted {
-    let is_static_call = is_fn_view(f);
+pub comptime fn register_private_fn_stub(f: FunctionDefinition) {
```
Newly introduced this function. Makes sense to go ahead and just register the stub here in the registry instead of in the `transform_private` function. This results in better encapsulation.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -19,24 +23,27 @@ comptime global FROM_FIELD: TypedExpr = {
 
 comptime global SERIALIZED_ARGS_NAME: Quoted = quote { serialized_args };
 
-pub comptime fn stub_fn(f: FunctionDefinition) -> Quoted {
```
The diff became a bit messy here as I dropped the `stub_fn` and put the register_* functions in its place.

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -1,10 +1,12 @@
-use crate::macros::utils::{
-    AsStrQuote, compute_fn_selector, is_fn_private, is_fn_public, is_fn_view,
-};
+// ## What are call interface stubs?
```
```suggestion
```

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/call_interface_stubs.nr:
```diff
@@ -1,10 +1,12 @@
+// Instead of manually serializing arguments and creating call interfaces, stubs allow natural syntax, e.g. for
+// enqueuing calls to public functions:
+//
+//   ExternalContract.at(address).some_method(arg1, arg2).enqueue()
```
Unrelated to this PR, but this is a perfect example of an instance when we'd use `//!` comments, which apply documentation to the parent item (the `mod call_interface_stubs`). So this is great, it'll just be a matter of leveraging the new docs tech once the time comes.

benesjan replied:
Cool. Was not aware of this type of comments.

It's a minor change so sneaked that in https://github.com/AztecProtocol/aztec-packages/pull/17840/commits/abe4d5c2d324c2c18a860175ebd5b8e907721c39

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/stub_registry.nr:
```diff
@@ -1,6 +1,7 @@
 // This file provides a registry for storing and retrieving function stubs generated by macros.
 // It maintains a global hashmap that maps modules to their generated stub functions,
 // allowing the macro system to collect and organize contract interface stubs.
+// See call_interface_stubs.nr for more details on what stubs are and how they are generated.
```
```suggestion
// See the call_interface_stubs module for more details on what stubs are and how they are generated.
```

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -92,7 +102,7 @@ pub(crate) comptime fn transform_private(f: FunctionDefinition) {
     // Inject the authwit check if the function is marked with #[authorize_once].
     let authorize_once_check = if fn_has_authorize_once(f) {
-        create_authorize_once_check(f)
+        create_authorize_once_check(f, true)
```
This boolean makes me a bit sad, but at least in the IDE the editor will display it as `is_private: true`, plus we'll change the `auth_once` macro anyway, plus this is all very internal, plus there are no enums in Noir.

But I'm still sad.

nventuro reviewed (approved):
Agree with your comments re. how this simplifies things. Also greatly appreciate how in the description you went over your thinking process, it made reading this much easier. Thanks!

benesjan committed:
refactor: cleaning up macros::functions::utils

This PR was quite a journey:

## Step 1

When trying to figure out how to rename the incorrectly named `get_fn_visibility` in a PR down the stack I realized that it's better to just drop it because:
- it was only used by `create_assert_correct_initializer_args`, `create_mark_as_initialized` and `create_init_check` functions which in turn where only called in `transfrom_private` and `transform_public`,
- these functions contained only 2 lines of code.

Because of this ^ it's better to just directly inline contents of these funcs into the `transform_*` functions.

When I finished this I checked where are the is `is_fn_private` and `is_fn_public` utils used which led me to step 2:

## Step 2
When checking when are the above mentioned functions used I realized that we use them in `create_authorize_once_check` and in the `stub_fn`. In both of these cases the code was just terrible because these functions are called from `transform_private` and `transform_public` functions hence the flow was:

`transform_private` --> `stub_fn` --> are we in private? --> if yes call `create_private_stub`

which I could replace with:

`transform_private` --> `create_private_stub`

So I dropped the `stub_fn` and just called the `create_private_stub` directly.

The similar was the case for `create_authorize_once_check` where I could simply pass in an `is_private` arg to the function and get rid of the `is_fn_private` dependency.

## Step 3
Once I did this I realized that I don't need the `EXTERNAL_REGISTRY` I introduced in the PR down the stack at all, removing a lot of the sad ugly complexity I needed to introduce there 🎉

## Future work
- The whole `noir-projects/aztec-nr/aztec/src/macros/functions` directory is very messy and I plan on move the code around into more logical places in the dir once this stack of PRs is merged.
- I think the `stub` name is non-descriptive enough so will try to come up with something better.

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
